spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,535 @@
+type MultihopProof struct {
+    KeyProofIndex uint32           // the index of the consensus state to prove the key/value on, normally 0
+    KeyProof *ProofData            // the key/value proof on the source chain (consensusState[0])
+    ConsensusProofs []*ProofData   // array of consensus proofs starting with proof of consensusState[0] on chain1
+    ConnectionProofs []*ProofData  // array of connection proofs starting with proof of conn[1,0] on chain1 (consensusState[1])
+}
+
+// GenerateMultihopProof generates proof of a key/value at the proofHeight on indexed chain (chain0).
+// Chains are provided in order from the sending (source) chain to the verifying chain.
+func GenerateMultihopProof(chains []*Chain, key string, value []byte, proofHeight exported.Height) *MultihopProof {
+
+    abortTransactionUnless(len(chains) > 2)
+
+    // generate and assign consensus, clientState, and connection proofs
+    var proof MultihopProof
+    proof.ConsensusProofs = GenerateConsensusProofs(chains)
+    proof.ConnectionProofs = GenerateConnectionProofs(chains)
+
+    chain0 := chains[0] // source chain
+    chain1 := chains[1] // next hop chain
+
+    // check a consensus state at proof height is present on chain 1
+    consensusStateAtProofHeight := chain1.GetConsensusStateAtHeight(chain0.ClientID, proofHeight)
+    abortTransactionUnless(consensusStateAtProofHeight != nil)
+
+    // query the key/value proof on the indexed chain at the proof height
+    keyProof, _ := chain0.QueryProofAtHeight(key, proofHeight)
+
+    // assign the key/value proof
+    proof.KeyProof = &ProofData{
+        Key:   nil,  // key to prove constructed during verification
+        Value: nil,  // proven values are constructed during verification (except for frozen client proofs)
+        Proof: keyProof,
+    }
+
+    return &proof
+}
+
+// GenerateFrozenClientProof generate a multihop proof of a frozen channel given a set of chains starting from the misbehaving chain.
+// Chains are provided in order from the misbehaving chain to the verifying chain.
+func GenerateFrozenClientProof(chains []*Chain, proofHeight exported.Height) *MultihopProof {
+
+    abortTransactionUnless(len(chains) > 2)
+
+    // generate and assign consensus, clientState, and connection proofs
+    var proof MultihopProof
+    proof.ConsensusProofs = GenerateConsensusProofs(chains)
+    proof.ConnectionProofs = GenerateConnectionProofs(chains)
+
+    chain0 := chains[0] // misbehaving chain
+    chain1 := chains[1] // next hop chain
+
+    // check a consensus state at proof height is present on chain 1
+    consensusStateAtProofHeight := chain1.GetConsensusStateAtHeight(chain0, proofHeight)
+    abortTransactionUnless(consensusStateAtProofHeight != nil)
+
+    // connectionEnd on misbehaving chain representing the counterparty connection on the next chain
+    counterpartyConnectionEnd := abortTransactionUnless(Unmarshal(proof.ConnectionProofs[0].Value))
+
```

Comment:
At first look this doesn't seem to work imho. Could we have a section going through an example? I think it would be useful as the proof generation in general is complex and for this case in particular.
So again let's take this example with connection IDs showing. Assume the client used by `bc` is frozen

`A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`

So we call:
1. `GenerateFrozenClientProof([B, A])` -> proofs will be sent to A
- In this case A will get connection proofs `[bc, a]` and will take the client of `bc`.

2. `GenerateFrozenClientProof([B, C, D])` -> proofs will be sent to D
- D will get proofs for `[cb, dc]` but should use the client of the counterparty of `cb`.

So 1. and 2. don't look the same wrt to proof collection logic. Same for verification in `chanCloseFrozen`.
Intuitively I was expecting some kind of indication of the "direction" being froze, not sure we can generalize for both 1 and 2. But I might be proven wrong.

Reply:
Yes, your assessment is correct. We made a decision to only allow the proof to work on the side with the frozen client (proof logic to A in step 1 from your example). We came to the same conclusion that it would be hard to generalize though this is certainly open for consideration.

The consequence of this choice is that in your example above only the channel on chain A could be forcefully frozen. In order for the channel on chain D to be frozen `dc` would need to be frozen (although in this example it would not be necessary since chain D would already see the frozen client state).

So in order to freeze both ends of the channel, evidence must be submitted to freeze the client state on both sides of the misbehaving chain. This way the proof logic is always the same, and the tradeoff is that the channel can only be frozen on the side of the channel path containing the frozen client.

Thoughts?

Reply:
_(A clarification that in my example those IDs were for connections but yes, we have similar for clientIDs)_ 

When we discussed this offline, I was asking the same question and feedback from @AdityaSripal was that we should close both ends. I agree with that, we should close both ends.
We cannot freeze a client on the other "side" as that client/ connection may be valid and used by other single or multi-hop channels.

I think we can make this function work if we add an explicit flag, e.g. `i` (incoming) to the last proof. By default proofs are `o`, i.e. outgoing.

_(Note that I flipped the proofs below as well, i.e. they can be verified from first to last)_
So for 2. the proof could be something like `[dc, cb, bc(i)]` vs for 1. `[a, bc]`
If the last proof is marked `i` then the client on the incoming/ counterparty connection end is frozen, i.e. `bc` must be for the counterparty of `cb`.

Reply:
I have one question with this approach. If `ClientStateC` on `ChainB` (represented as `bc`) is frozen, would that be provable on `ConsensusStateB` on `ChainC`? Once the client is frozen there will not be any more consensus updates so would `ConsensusStateB` be frozen just prior to the freeze or would that state be included in the state?

That said, I went back and looked at the implementation for the ChanCloseFrozen proof, and I actually think the existing logic can support channel freezing on both ends of the channel with very small changes by using the new `KeyProofIndex` field of the MultihopProof struct to also indicate the directionality of the proof. Going back to your example above:

`A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`

1. For A, its `multihopConnectionEnd` is `bc`. The ClientID for ClientBC should be `multihopConnectionEnd.Counterparty.ClientD` and we can use this logic to determine the client state key if the `KeyProofIndex != 0`

2. For D, its `multihopConnectionEnd` is `cb`. The ClientID for ClientBC should be `multihopConnectionEnd.ClientID` and we can use this to determine the key for the client state if `KeyProofIndex == 0`

`KeyProofIndex` is a new field introduced to keep the multi-hop verification logic the same while allowing verification of keys within intermediate consensus states in the proof chain.

Reply:
> I have one question with this approach. If `ClientStateC` on `ChainB` (represented as `bc`) is frozen, would that be provable on `ConsensusStateB` on `ChainC`? Once the client is frozen there will not be any more consensus updates so would `ConsensusStateB` be frozen just prior to the freeze or would that state be included in the state?
> 

Not sure I understand. If the client for `bc` is frozen at height `hb` we can still update the client for `cb` on C with a consensus state of chain B at a height `hb', hb'>hb`. Then we can query B's client for `bc` with proof at height `hb'`.

> That said, I went back and looked at the implementation for the ChanCloseFrozen proof, and I actually think the existing logic can support channel freezing on both ends of the channel with very small changes by using the new `KeyProofIndex` field of the MultihopProof struct to also indicate the directionality of the proof. Going back to your example above:
> 
> `A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`
> 
> 1. For A, its `multihopConnectionEnd` is `bc`. The ClientID for ClientBC should be `multihopConnectionEnd.Counterparty.ClientD` and we can use this logic to determine the client state key if the `KeyProofIndex != 0`

shouldn't be `multihopConnectionEnd.ClientID` in this case? and ``KeyProofIndex == 0`
> 2. For D, its `multihopConnectionEnd` is `cb`. The ClientID for ClientBC should be `multihopConnectionEnd.ClientID` and we can use this to determine the key for the client state if `KeyProofIndex == 0`
> 

shouldn't be `multihopConnectionEnd.Counterparty.ClientD` and `KeyProofIndex != 0`

> `KeyProofIndex` is a new field introduced to keep the multi-hop verification logic the same while allowing verification of keys within intermediate consensus states in the proof chain.

So iiuc, `KeyProofIndex` is non-zero in case 2 of proofs for closing channels due to on-path client being frozen.

Reply:
I was thinking about this when reviewing and got stuck exactly here...
> Here, I would say that a relayer should first freeze the clients of C on the direct counterparties. Namely, (bc) and (dc) should be frozen.

hermes currently submits evidence only once a client is updated. If `bc` client is updated (by some other relayer) hermes compares the header used in the update with the header it gets from the C full node it connects to. Then it submits `MsgMisbehaviour` but to `bc` only.

The relayer could try to freeze all clients of C as you point out. But there is no guarantee that this works, in the example, the header that updated `bc` might not be verifiable on the `dc` client because the clients may have different security parameters (specifically the trusting level). So in theory we have a small probability channel on E will not be closed because we cannot freeze `dc`. 
 
Can we live with this in practice? It would simplify the spec quite a bit, we could eliminate `KeyProofIndex`, etc.